pyls/_utils.py

```diff
@@ -9,21 +9,49 @@
 
 FIRST_CAP_RE = re.compile('(.)([A-Z][a-z]+)')
 ALL_CAP_RE = re.compile('([a-z0-9])([A-Z])')
+KWD_MARK = object()
 
 
-def debounce(interval_s):
+def debounce(interval_s, keys=None):
     """Debounce calls to this function until interval_s seconds have passed."""
     def wrapper(func):
+        arg_cache = {}
+        lock = threading.Lock()
+
+        @functools.wraps(func)
+        def cleanup(*args, **kwargs):
+            input_hash = _hash_input(keys, *args, **kwargs)
+            with lock:
+                del arg_cache[input_hash]
+            return func(*args, **kwargs)
+
         @functools.wraps(func)
         def debounced(*args, **kwargs):
-            if hasattr(debounced, '_timer'):
-                debounced._timer.cancel()
-            debounced._timer = threading.Timer(interval_s, func, args, kwargs)
-            debounced._timer.start()
+            input_hash = _hash_input(keys, *args, **kwargs)
+            with lock:
+                if input_hash in arg_cache:
+                    arg_cache[input_hash].cancel()
+                timer = threading.Timer(interval_s, cleanup, args, kwargs)
+                arg_cache[input_hash] = timer
+                timer.start()
         return debounced
     return wrapper
 
 
+def _hash_input(keys, *args, **kwargs):
```

Comment:
Get rid of this and just use `kwargs[key]` - can make it more complicate if/when you need it